fix: update Beacon api impl to enforce unique array

mkermani144 opened the pull request:
**Motivation**

Make sure array inputs to some Beacon APIs are unique, based on [Beacon APIs spec](https://github.com/ethereum/beacon-APIs).

**Description**

This change enhances the robustness of the API by ensuring that duplicate entries are not processed, aligning with schema definitions from Beacon API spec.

Closes #5271

**Steps to test or reproduce**
- Run a dev environment via `lodestar dev`
- Test the API with duplicate input (=query, body, etc.)
- Assert that a response with 400 status code is returned

mkermani144 committed:
fix: update Beacon api impl to enforce unique array

This change enhances the robustness of the API by ensuring that
duplicate entries are not processed, aligning with schema definitions
from Beacon API spec.

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/index.ts:
```diff
@@ -85,6 +86,9 @@ export function getBeaconStateApi({
     },
 
     async getStateValidators({stateId, validatorIds = [], statuses = []}) {
+      ensureUniqueItemsOrThrow(validatorIds, "Duplicate validator IDs provided");
+      ensureUniqueItemsOrThrow(statuses, "Duplicate statuses provided");
```
this only needs to be checked if `length > 0` which we do below for each, should move those there

mkermani144 replied:
Moved to condition blocks ✅

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/index.ts:
```diff
@@ -15,6 +15,7 @@ import {getValidatorStatus} from "@lodestar/types";
 import {fromHex} from "@lodestar/utils";
 import {ApiError} from "../../errors.js";
 import {ApiModules} from "../../types.js";
+import ensureUniqueItemsOrThrow from "../../utils.js";
```
we don't use default exports usually

mkermani144 replied:
Changed to named export ✅

nflaig commented on packages/beacon-node/src/api/impl/utils.ts:
```diff
@@ -0,0 +1,22 @@
+ * @param message - The message to put in the ApiError if the array contains
+ * duplicates.
+ */
+const ensureUniqueItemsOrThrow = (array: unknown[] | undefined, message: string) => {
```
```suggestion
export function assertUniqueItems(array: unknown[] | undefined, message: string) {
```

mkermani144 replied:
✅

nflaig commented on packages/beacon-node/src/api/impl/utils.ts:
```diff
@@ -0,0 +1,22 @@
+    return;
+  }
+
+  const containsDuplicates = new Set(array).size !== array.length;
```
it would be kinda nice to be able to return the duplicate values are part of the error so it's easier for consumers to figure out what went wrong

mkermani144 replied:
Switched to a diff-like method to find the duplicates, and introduced a new error type which concats the duplicates at the end of the error message ✅

nflaig commented on packages/beacon-node/src/api/impl/utils.ts:
please try to follow current code structure here, ie. how the function is defined and exported does not align with our code style

mkermani144 replied:
- Switched to named exports
- Switched to function declaration (instead of arrow)
- Added return type

Let me know if I'm missing any other convention.

mkermani144 committed:
feat: introduce a new error type for better ux

This update adds a new error class, ApiInputDuplicateItemsError, to handle cases where API inputs contain duplicate items. The ensureUniqueItemsOrThrow function has been modified to utilize this new error class, improving error message and clarity in API responses. Additionally, relevant tests have been updated to reflect these changes.

Some code styles are also improved.

nflaig commented on packages/beacon-node/src/api/impl/errors.ts:
```diff
@@ -47,4 +47,17 @@ export class IndexedError extends ApiError {
+ * Error thrown when an array input to an API (e.g. its body, params, etc.)
+ * contains duplicate items.
+ */
+export class ApiInputDuplicateItemsError<T> extends ApiError {
```
why do we need a custom error for this?

mkermani144 replied:
I overlooked its redundancy. I first thought this will be reused for each validation, but after extracting the validation logic into a `ensureUniqueItemsOrThrow` function, it is of no use. Removed it.

mkermani144 committed:
chore: remove redundant ApiInputDuplicateItemsError

This error is only used once, and has no use case.

mkermani144 commented:
@nflaig Any idea why sim tests constantly failing? Is it a problem caused by my changes?

nflaig commented on packages/beacon-node/src/api/impl/utils.ts:
```diff
@@ -0,0 +1,25 @@
+  }, []);
+
+  if (duplicateItems.length) {
+    throw new ApiError(400, `${message} (Duplicate Items: ${duplicateItems.join(", ")})`);
```
message seems redundant, what about

```suggestion
    throw new ApiError(400, `${message}: ${duplicateItems.join(", ")}`);
```

nflaig commented on packages/beacon-node/src/api/impl/utils.ts:
```diff
@@ -0,0 +1,25 @@
+ * @param message - The message to put in the ApiError if the array contains
+ * duplicates.
+ */
+export function ensureUniqueItemsOrThrow(array: unknown[] | undefined, message: string): void {
```
can we rename this please (as suggested earlier)

```suggestion
export function assertUniqueItems(array: unknown[] | undefined, message: string): void {
```

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/index.ts:
```diff
@@ -231,6 +238,8 @@ export function getBeaconStateApi({
     },
 
     async postStateValidatorBalances(args, context) {
+      ensureUniqueItemsOrThrow(args.validatorIds, "Duplicate validator IDs provided");
```
this should be enforced in `this.getStateValidatorBalances`, right now it's checking for unique items twice

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/index.ts:
```diff
@@ -194,6 +199,8 @@ export function getBeaconStateApi({
     },
 
     async getStateValidatorBalances({stateId, validatorIds = []}) {
+      ensureUniqueItemsOrThrow(validatorIds, "Duplicate validator IDs provided");
```
this can be checked after `if (validatorIds.length)` at line 206

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/index.ts:
```diff
@@ -142,6 +145,8 @@ export function getBeaconStateApi({
     },
 
     async postStateValidatorIdentities({stateId, validatorIds = []}) {
+      ensureUniqueItemsOrThrow(validatorIds, "Duplicate validator IDs provided");
```
can be moved to line 156

nflaig commented:
> @nflaig Any idea why sim tests constantly failing? Is it a problem caused by my changes?

I ran the CI again, will check if it's still failing, I don't see why changes in the PR would cause issues though

mkermani144 committed:
chore: improve error message for duplicate items in API input

mkermani144 committed:
chore: rename ensureUniqueItemsOrThrow to assertUniqueItems

mkermani144 committed:
chore: improve `assertUniqueItems` call sites

mkermani144 committed:
chore: fix linting errors

nflaig committed:
Merge branch 'unstable' into fix/enforce-api-input-uniqueness

nflaig committed:
Review PR

nflaig reviewed (approved):
LGTM - thanks @mkermani144

wemeetagain commented:
:tada: This PR is included in [v1.31.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.31.0) :tada:
